java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/MessageReceiver.java

```diff
@@ -300,16 +300,16 @@ public void setReceiveTimeout(final Duration value)
 				@Override
 				public void onEvent()
 				{
-					if (receiveLink.getLocalState() == EndpointState.CLOSED || receiveLink.getRemoteState() == EndpointState.CLOSED)
-					{
-						createReceiveLink();
-					}
-
 					final List<Message> messages = receiveCore(maxMessageCount);
 					if (messages != null)
 						onReceive.complete(messages);
 					else
 						pendingReceives.offer(new ReceiveWorkItem(onReceive, receiveTimeout, maxMessageCount));
+
+					if (receiveLink.getLocalState() == EndpointState.CLOSED || receiveLink.getRemoteState() == EndpointState.CLOSED)
+					{
+						createReceiveLink();
+					}
```

Comment:
good comment @hmlam. I belv. a strict contract needs to be incorporated in couple other places. filed an issue to track this: #244